usaspending_api/common/helpers/s3_helpers.py

```diff
@@ -0,0 +1,52 @@
+import boto3
+import io
+import logging
+import re
+
+from typing import Optional, List
+
+from usaspending_api import settings
+
+logger = logging.getLogger("script")
+
+
+def access_s3_object_list(
+    bucket_name: str, regex_pattern: Optional[str] = None
+) -> List["boto3.resources.factory.s3.ObjectSummary"]:
+    """Find all S3 objects in provided bucket.
+
+    If regex is passed, only keys which match the regex are returned
+    """
+
+    logger.info("Gathering all deleted transactions from S3")
+    try:
+        bucket = get_s3_bucket(bucket_name=bucket_name)
+        bucket_objects = list(bucket.objects.all())
+    except Exception:
+        msg = (
+            "Verify settings.USASPENDING_AWS_REGION and settings.DELETED_TRANSACTION_JOURNAL_FILES are correct "
+            "or env variables: USASPENDING_AWS_REGION and DELETED_TRANSACTION_JOURNAL_FILES are set"
+        )
+        logger.exception(msg)
+        bucket_objects = None
+
+    if regex_pattern and bucket_objects:
+        bucket_objects = [obj for obj in bucket_objects if re.search(regex_pattern, obj.key)]
+
+    return bucket_objects
+
+
+def get_s3_bucket(
+    bucket_name: str, region_name: str = settings.USASPENDING_AWS_REGION
+) -> "boto3.resources.factory.s3.Instance":
+    s3 = boto3.resource("s3", region_name=region_name)
+    return s3.Bucket(bucket_name)
+
+
+def access_s3_object(bucket_name: str, obj: "boto3.resources.factory.s3.ObjectSummary") -> io.BytesIO:
+    """Return the Bytes of an S3 object"""
+    bucket = get_s3_bucket(bucket_name=bucket_name)
+    data = io.BytesIO()
+    bucket.download_fileobj(obj.key, data)
+    data.seek(0)  # Like rewinding a VCR cassette
```

Comment:
If I understand this functionality correctly, this wouldn't be thread safe. Is that a concern?

Reply:
Perhaps you can elaborate more on "this" (entire function? `data.seek(0)`?)

Are any of the current management commands which would use this multi-threaded?

Reply:
specifically the use of `data.seek(0)`, if this means the S3 object has some kind of global state that needs to be reset, makes `access_s3_object` not thread safe. I don't know if this is a problem in practice, but maybe might be an issue if multiple management commands are being run in Jenkins, and this is connecting to some state outside the app. The [base documentation](https://docs.python.org/3/library/io.html) of io seems mostly about file reading, so I wasn't able to find a quick answer. 
If this isn't a concern, and we know that io is just storing a pointer to the last line locally (which is what I would expect this kind of library to be doing), is there any need to rewind a cassette that we're just going to throw away?

Reply:
The s3 object data is being downloaded using `. download_fileobj()` into `data` (a byteIO object) and the seek on the python variable holding the data is immediately after, the code isn't manupulating anything in AWS nor is it expecting external state other than an existing s3 bucket name and object name. I don't see a risk of something occurring between those two lines even if called multiple times by a multithreaded script.